server/lib/util.py

```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
   if is_json:
     response.headers['Content-Type'] = 'application/json'
   return response
+
+
+def fetch_highest_coverage(parent_entity: str,
+                           child_type: str,
+                           variables: List[str],
+                           all_facets: bool,
+                           facet_ids: List[str] = None):
+  """
+  Fetches the latest available data with the best coverage for the given a
+  parent entity, child type, variables, and facets
+
+  Response format:
+  {
+    "facets": {
+      <facet_id>: {<facet object>}
+    },
+    "data": {
+      <var_dcid>: {
+        <entity_dcid>: {
+          <observation point(s)>
+        }
+      }
+    }
+  }
+  """
+  # Heuristic for fetching "latest data with highest coverage":
+  # Choose the date with the most data coverage from either:
+  # (1) last 5 observation dates
+  # (2) 5 years from the most recent observation date
+  # whichever set is greater
+  FETCH_HIGHEST_COVERAGE_FROM_DATES = 5
+  FETCH_HIGHEST_COVERAGE_FROM_YEARS = 5
+  point_responses = []
+  series_dates_response = dc.get_series_dates(parent_entity, child_type,
+                                              variables)
+  facet_ids_set = set(facet_ids or [])
+  for variable_date_item in series_dates_response['datesByVariable']:
+    # Each variable_date_item contains the observation counts by date
+    variable = variable_date_item['variable']
+    # Get observation dates in descending order, and filter out dates in the
+    # future (to exclude erroneous data)
+    next_year = str(date.today().year + 1)
+    descending_observation_dates = [
+        observation_date for observation_date in list(
+            reversed(variable_date_item.get('observationDates', [])))
+        if observation_date['date'] < next_year
+    ]
+    if len(descending_observation_dates) == 0:
+      continue
+    # Heuristic to fetch the "FETCH_HIGHEST_COVERAGE_FROM_DATES" most recent
+    # observation dates or observation dates going back
+    # "FETCH_HIGHEST_COVERAGE_FROM_YEARS" years, whichever is greater
+    cutoff_year = str(date.today().year - FETCH_HIGHEST_COVERAGE_FROM_YEARS)
+    latest_observation_dates_from_year = [
+        o for o in descending_observation_dates if o['date'] > cutoff_year
+    ]
+    latest_observation_dates = descending_observation_dates[:
+                                                            FETCH_HIGHEST_COVERAGE_FROM_DATES]
+    observation_dates = latest_observation_dates_from_year if len(
```

Comment:
why comparing the length not the latest date from the two list?

Reply:
The latest date from the two lists should always be the same (the most recent year). This logic just checks how far to walk backwards in time from the latest date.